cognite/client/data_classes/_base.py

```diff
@@ -268,8 +268,11 @@ def _build_id_mappings(self) -> None:
         self._id_to_item, self._external_id_to_item, self._instance_id_to_item = {}, {}, {}
         if not self.data:
             return
-        if hasattr(self.data[0], "external_id"):
-            self._external_id_to_item = {item.external_id: item for item in self.data if item.external_id is not None}
+        self._external_id_to_item = {
+            item.external_id: item
+            for item in self.data
+            if hasattr(item, "external_id") and item.external_id is not None
+        }
```

Comment:
The principal breaks this assumption. There are service principal and user principals, were a service principal has an external ID while an internal does not.